start on scala integration tutorial (#16)
sifive/block-pio-sifive: README.md

# Overview

> This branch represents the Parallel IO block and loopback VIP in their

Reviewer: Let's spell out Verification IP the first time it is used, with the acronym in parentheses, since it may not be a term that everyone is familiar with. Also, let's put PIO in parentheses this first time to similarly introduce an acronym.

```suggestion
This branch represents the Parallel IO (PIO) block and loopback VIP (Verification IP) in their
```

preonboarded state. This README is a step-by-step tutorial of how to onboard
the PIO block.

The PIO block communicates to the CPU through an AXI4 interface. This interface
is used to read from and drive the `odata`, `oenable`, and `idata` control
registers that map to top-level ports. To integrate this block we need make the
odata, oenable, and idata ports available at the top-level of design and

> connect the AXI4 interface to Rocket's periphery bus.

Reviewer: I think it would be more correct to call it the "Test Socket's peripheral bus", not "Rocket's". Rocket is just the name of the core, which excludes the interconnect. I could also see calling it the "Rocket core subsystem's peripheral bus", if you think the Test Socket is too confusing of a concept.

> There is also an accompanying vip for testing our block. The loopback vip needs

Reviewer: ```suggestion
There is also an accompanying VIP for testing our block. The loopback VIP needs
```

> to be instantiated in the testharness and be connected to the odata, oenable,

Reviewer: ```suggestion
to be instantiated in the test harness and be connected to the odata, oenable,
```

```bash
duh validate pio.json5
```

## Scala integration
In order to integrate our IP block into the Craft framework we need create a
Scala wrapper for our block that is parametrizeable and can be automatically

> attached to an SoC. This section describes how to generate boiler-plate Scala

Reviewer: ```suggestion
attached to an SoC. This section describes how to generate boilerplate Scala
```

> attached to an SoC. This section describes how to generate boiler-plate Scala

Reviewer: Aside from grammatical comments, I think it might also be worth mentioning that the "can be automatically attached to an SoC" part is done by defining some functions that ultimately connect Diplomatic nodes of the block to the SoC.

code from a DUH document and what modifications need to be made in order to
fully integrate the PIO block.

### Generating Scala code with DUH

> The best way to onboard a verilog IP block is to use `duh-export-scala` to

Reviewer: ```suggestion
The best way to onboard a Verilog IP block is to use `duh-export-scala` to
```

generate an extensible Scala wrapper. We follow the convention of putting Scala
code in the `craft/${design}/src` directory. Run the following commands to

> generate the scala wrapper files.

Reviewer: ```suggestion
generate the Scala wrapper files.
```

```bash
duh-export-scala pio.json5 -o craft/pio/src
duh-export-scala loopback.json5 -o craft/loopback/src
```

This will create `${name}.scala` and `${name}-base.scala` files in the
`craft/pio/src` and `craft/loopback/src` directory. The `*-base.scala` files
contain base class definitions and should not be edited. Any changes to
`*-base.scala` files will be overwritten whenever `duh-export-scala` is rerun.

The `${name}.scala` files contain classes that extend the base classes defined
in `${name}-base.scala` and can be modified to override or augment the
base class functionality.

For a more detailed explanation of the generated scala see the
[duh-scala README](https://github.com/sifive/duh-scala#blackbox-wrapper-api).

### Extending the `duh` generated base classes
By default, ports that are not associated with a bus interface are left
unconnected. We need to add the following changes we to the generated base
classes to create top-level ports for the `oenable`, `odata`, and `idata`
signals of the PIO block and connect them to the loopback VIP.

> First we need to import classes from our loopback scala wrapper into

Reviewer: ```suggestion
First we need to import classes from our loopback Scala wrapper into
```

Copy the following lines into the body of the `NpioTop`class in `pio.scala`.
```scala
  // route the ports of the black box to this sink
  val ioBridgeSink = BundleBridgeSink[pioBlackBoxIO]()
  ioBridgeSink := imp.ioBridgeSource

  // create a new ports for odata, oenable, and idata
  val ioBridgeSource = BundleBridgeSource(() => new NpioTopIO(c.blackbox.pioWidth))

  // logic to connect ioBridgeSink/Source nodes
  override lazy val module = new LazyModuleImp(this) {

    // connect the clock and negedge reset to the default clock and reset
    ioBridgeSink.bundle.clk     := clock.asUInt
    ioBridgeSink.bundle.reset_n := !(reset.toBool)

    // connect ioBridge source and sink
    ioBridgeSource.bundle.odata   := ioBridgeSink.bundle.odata
    ioBridgeSource.bundle.oenable := ioBridgeSink.bundle.oenable
    ioBridgeSink.bundle.idata     := ioBridgeSource.bundle.idata
  }
```

An important artifact of the Scala generator is the Object Model JSON. This
JSON file contains important information about the generated instance of the
design. To ensure our IP block is fully integrated with our tools we need to
populate the section of Object Model that describes our PIO block.

> By default the generated scala will populate the `OMMemoryRegion`s of attached

Reviewer: ```suggestion
By default the generated Scala will populate the `OMMemoryRegion`s of attached
```

bus interfaces and `OMInterrupts` of our block's interrupt bus interfaces. We
need to add a field describing the `pioWidth` parameter of our block and also
associate an `OMRegisterMap` with the AXI4 memory region.

First we need to define a case class containing fields that we want to add to
the Object Model. Add the following case class definition to `pio.scala`
```scala
case class OMPIO(width: Int)
```

Then override the `userOM` and `getOMMemoryRegions` methods of `NpioTop` to add

> our custom Object Model and add register maps to our blocks memory regions

Reviewer: ```suggestion
our custom Object Model and add register maps to our block's memory regions
```

respectively.
```scala
  // add in custom fields to the Object Model entry for this block
  override val userOM: OMPIO = OMPIO(c.blackbox.pioWidth)

  // associate register maps with memory regions in Object model
  override def getOMMemoryRegions(resourceBindings: ResourceBindings) = {
    super.getOMMemoryRegions(resourceBindings).zip(omRegisterMaps).map { case (memRegion, regmap) =>
      memRegion.copy(registerMap = Some(regmap))
    }
  }
```

Finally, we need to edit `attach` method in the `NpioTop` companion object to

> instantiate the loopback VIP in the testharness and connect it to the `oenable`,

Reviewer: ```suggestion
instantiate the loopback VIP in the test harness and connect it to the `oenable`,
```

`odata`, and `idata` ports of the PIO block. `bap.testHarness` is a `LazyScope`
which means that when you apply it to a code block
(i.e. `bap.testHarness { $my_code_block}`) that code block will execute as if

> it was part of the `bap.testHarness` module. Replace the line

Reviewer: ```suggestion
it were part of the `bap.testHarness` module. Replace the line
```

> `// User code here` in `attach` method in the `NpioTop` companion object with

Reviewer: ```suggestion
`// User code here` in the `attach` method of the `NpioTop` companion object with
```
